### PR TITLE
Add BatchNorm2d

### PR DESCRIPTION
**Pull request by svtdanny**

### CODE REVIEW

File: wrappers/python/nntile/tensor.py [hunk 1]
```diff
@@ -11,811 +11,6 @@
...
-# Wrapper for multiprecision transpose
-def transpose_async(alpha: float, src: Tensor, dst: Tensor, ndim: int) -> None:
-    if type(src) is not type(dst):
-        raise TypeError
-    if type(src) is core_tensor.Tensor_fp32:
-        core_tensor.transpose_async_fp32(alpha, src, dst, ndim)
-    elif type(src) is core_tensor.Tensor_fp64:
-        core_tensor.transpose_async_fp64(alpha, src, dst, ndim)
-    else:
-        raise TypeError
```
Comment (daskol) [inline wrappers/python/nntile/tensor.py:820]: Do not rework anything in feature PR. Prepend this PR with a PR where you rework this module.

File: wrappers/python/nntile/utils/constructors.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+np2nnt_type_mapping = {
+    np.float32: Tensor_fp32,
+    np.float64: Tensor_fp64
+}
+
+
+
+def tensor_from_numpy(A: np.array, basetile_shape: tuple = None, mpi_distr = [0], next_tag = 0):
```
Comment (daskol) [inline wrappers/python/nntile/utils/constructors.py:31]: It better to name it in NumPy-like way. Something like `astensor` or just `tensor`.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+def batch_norm_test_suite():
+    for _iter in range(4):
+        test_batchnorm(BatchNormTestParams([4,5,10,10], np.float32))
+        test_batchnorm(BatchNormTestParams([4,5,10,10], np.float64))
+
+        test_batchnorm(BatchNormTestParams([1,1,10,10], np.float32))
+
+if __name__ == "__main__":
+    batch_norm_test_suite()
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:97]: Firstly, use `pytest`.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import nntile.utils
+import nntile.utils.constructors
+import numpy as np
+import torch
+import torch.nn as nn
+from dataclasses import dataclass
+
+config = nntile.starpu.Config(1, 0, 0)
+
+nntile.starpu.init()
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:26]: Secondly, wrap runtime initialization in module or function scoped fixture.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+dtypes = [np.float32, np.float64]
+
+@dataclass
+class BatchNormTestParams:
+    shape: list
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:31]: Usually, shape is inmmutable so it should be a tuple.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    dtype: np.dtype
+
+    atol: float = 1e-05
+
+    eps: float = 1e-05
+    redux: bool = False
+
+def generate_input(params: BatchNormTestParams):
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:39]: Since every parameter except `shape` is constant, pass them as kwargs and parameterize test with `pytest.mark.parameterize`.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    input_np = np.random.rand(*params.shape).astype(params.dtype)
+    output_grad_np = np.random.rand(*params.shape).astype(params.dtype)
+
+    input_torch = torch.tensor(input_np, requires_grad=True)
+    output_grad_torch = torch.tensor(output_grad_np, requires_grad=False)
+
+    # input_np = np.asfortranarray(input_np)
+    # output_grad_np = np.asfortranarray(output_grad_np)
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:47]: Remove dead code.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    input_grad_nnt = nntile.tensor.tensor_from_numpy(np.zeros(params.shape).astype(params.dtype))
+    input_moment = nntile.tensor.TensorMoments(input_nnt, input_grad_nnt, grad_required=True)
+    
+    output_grad_nnt = nntile.tensor.tensor_from_numpy(output_grad_np)
+
+    return (input_moment, output_grad_nnt), (input_torch, output_grad_torch)
+
+
+def test_batchnorm(params: BatchNormTestParams):
+    (input_moment, output_grad_nnt), (input_torch, output_grad_torch) = generate_input(params)
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:59]: Ah, I see. Generate input should be used as a fixture.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    nntile_layer.grad = output_grad_nnt
+    nntile_layer.backward_async()
+
+    # test forward
+    success = torch.allclose(
+        torch.tensor(nntile.tensor.tensor_to_numpy(nntile_layer.x_res)),
+        o,
+        atol=params.atol
+    )
+    assert success, "Error in forward for params: " + str(params)
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:79]: Firstly, it better to use string interpolation. Secondly, use `numpy.testing.assert_*` routines.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    output_grad_nnt = nntile.tensor.tensor_from_numpy(output_grad_np)
+
+    return (input_moment, output_grad_nnt), (input_torch, output_grad_torch)
+
+
+def test_batchnorm(params: BatchNormTestParams):
+    (input_moment, output_grad_nnt), (input_torch, output_grad_torch) = generate_input(params)
+    
+    # torch forward/backward
+    o = nn.BatchNorm2d(params.shape[1], dtype=input_torch.dtype, affine=False)(input_torch)
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:62]: Extremely bad variable naming. It is okay use one letter names in short expression like list comprehension or loops.

File: wrappers/python/tests/layer/test_batch_norm.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    input_nnt = nntile.tensor.tensor_from_numpy(input_np)
+    input_grad_nnt = nntile.tensor.tensor_from_numpy(np.zeros(params.shape).astype(params.dtype))
+    input_moment = nntile.tensor.TensorMoments(input_nnt, input_grad_nnt, grad_required=True)
+    
+    output_grad_nnt = nntile.tensor.tensor_from_numpy(output_grad_np)
+
+    return (input_moment, output_grad_nnt), (input_torch, output_grad_torch)
+
+
+def test_batchnorm(params: BatchNormTestParams):
```
Comment (daskol) [inline wrappers/python/tests/layer/test_batch_norm.py:58]: It's better to split test on two. The first one checks the forward pass. The second one checks the backward pass. You can organize them as a class to make them scoped.

```python
@pytest.parametrize('shape', [...])
class TestBatchNorm:
  def test_forward(self, ...):  ...
  def test_backward(self, ...):  ...
```

File: wrappers/python/nntile/layer/batch_norm.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    prod_fiber_async, prod_fiber3_async,
+    add_async,
+
+    zeros, ones
+)
+
+from nntile.layer.base_layer import BaseLayer
+
+class BatchNorm2d(BaseLayer):
+    def __init__(self, x: TensorMoments, weight: Tensor, bias: Tensor, x_res_tm: TensorMoments, x_mean: Tensor, inv_std: Tensor, eps: float = 1e-05, redux: bool = False):
```
Comment (daskol) [inline wrappers/python/nntile/layer/batch_norm.py:32]: Try to fit 80 columns.

File: wrappers/python/nntile/layer/batch_norm.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        self.eps = eps**0.5 # we will square it in forward (implementation constraints)
+
+        # TODO: add usage of learnable transforms 
+        self.weight = weight
+        self.bias = bias
+
+        self.x_res_tm = x_res_tm
+        self.x_res = self.x_res_tm.value
+
+        self.x_normalized_copy = zeros(self.x_res.shape, dtype=type(x.value))
```
Comment (daskol) [inline wrappers/python/nntile/layer/batch_norm.py:44]: What is `type(value)`? I guess, `x.value` is either a tensor or `None`. Then you use a type or `None` as a dtype.

File: wrappers/python/nntile/utils/constructors.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    A_traits = TensorTraits(A.shape, basetile_shape if basetile_shape else A.shape)
+    
+    A_value = np2nnt_type_mapping[A.type](A_traits, mpi_distr, next_tag)
+
+    A_value.from_array(A)   
+    return A_value
+
+
+def tensor_to_numpy(tensor_nnt):
+    np_res = np.zeros(tensor_nnt.shape, order='F').astype(np2nnt_type_mapping[type(tensor_nnt)])
```
Comment (daskol) [inline wrappers/python/nntile/utils/constructors.py:41]: Check `np.zeros` signature. It allows to do something like the following in single call.

```python
dtype = np2nnt_type_mapping[type(tensor_nnt)]
np_res = np.zeros(tensor_nnt.shape, dtype=dtype, order='F')
```

Also, you can avoid the second pass over NumPy-array due to initialization with `np.empty`.

File: wrappers/python/nntile/utils/constructors.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    tensor_nnt.to_array(np_res)
+    return np_res
+
+def zeros(shape, dtype=Tensor_fp32):
+    np_dtype = nnt2np_type_mapping[dtype]
+    return tensor_from_numpy(np.zeros(shape).astype(np_dtype))
```
Comment (daskol) [inline wrappers/python/nntile/utils/constructors.py:47]: What is it? Create array in order to create array?

File: wrappers/python/nntile/layer/batch_norm.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+        self.tmp_buff_full = zeros(self.x.shape, dtype=type(x.value))
+
+        self.grad_x_inv_denominator = zeros(self.x.shape, dtype=type(x.value))
+        
+        self.tmp_buff_channels = zeros([self.x.shape[1]], dtype=type(x.value))
+        self.tmp_buff_channels_internal = zeros([self.x.shape[1]], dtype=type(x.value))
+
+    @staticmethod
+    def generate_simple(x, eps: float = 1e-05, redux=False):
```
Comment (daskol) [inline wrappers/python/nntile/layer/batch_norm.py:67]: In general, it is better to use `classmethod`.

**Review by daskol — CHANGES_REQUESTED** [review]

File: wrappers/python/nntile/layer/batch_norm.py (hunk 13 above)
Comment (svtdanny) [reply wrappers/python/nntile/layer/batch_norm.py:44]: Not a good idea to pass None to batchnorm. It will crash, it seems to be expected. 
We can add assert over not None - what do you think?

File: wrappers/python/nntile/utils/constructors.py (hunk 15 above)
Comment (daskol) [reply wrappers/python/nntile/utils/constructors.py:47]: Addition input validation is required.

```python
import math
from typing import Sequence, TypeAlias

import numpy as np

ArrayLike: TypeAlias = Tensor | np.ndarray | ...


def array(object, tiling: Sequence[int], dtype=None, ...) -> Tensor:
    match object:
        case np.ndarray():
            tmp = object.astype(dtype=dtype, order='F', copy=False)
            arr = empty_like(tmp, tiling=tiling)
            arr.from_array(tmp)
        case _:
            raise NotImplementedError(
                f'No coersion from {type(object)} to tiled array is '
                'implemented at the momment.')


def empty(shape: Sequence[int], tiling: Sequence[int],
          dtype=None) -> Tensor:
    dtype = dtype or np.float32
    match dtype:
        case np.float32:
            tensor_ty = Tensor_fp32
        case np.float64:
            tensor_ty = Tensor_fp64
        case _:
            raise NotImplementedError(
                f'Tensor does not support {dtype} dtype.')

    bounds = ()
    for dim, tile_dim in zip(shape, tiling):
        if dim % tile_dim != 0:
            raise NotImplementedError(
                'Tile shape does not fit tensor shape evenly.')
        bounds += (dim // tile_dim,)
    notiles = math.prod(bounds)

    # TODO(@daskol): Human-friendly interface for device placement.
    placement = [0] * notiles
    traits = TensorTraits(shape, tiling)
    return tensor_ty(traits, placement, 0)


def empty_like(arr: ArrayLike, tiling: Sequence[int] | None = None,
               dtype=None) -> Tensor:
    if tiling is None:
        match arr:
            case np.ndarray():
                tiling = arr.shape
            case Tensor():
                tiling = arr.tiling
            case _:
                raise TypeError(
                    f'Unexpected array-like value of {type(arr)} type.')
    return empty(arr.shape, tiling, arr.dtype)


def full(shape: Sequence[int], tiling: Sequence[int], fill_value, dtype=None):
    arr = empty(shape, tiling, dtype)
    fill_value = arr.dtype(fill_value)
    fill_async(fill_value, arr)
    return arr


def ones(shape: Sequence[int], tiling: Sequence[int], dtype=None):
    return full(shape, tiling, 1, dtype)


def zeros(shape: Sequence[int], tiling: Sequence[int], dtype=None):
    return full(shape, tiling, 0, dtype)
```

File: wrappers/python/nntile/layer/batch_norm.py (hunk 13 above)
Comment (daskol) [reply wrappers/python/nntile/layer/batch_norm.py:44]: I belive that the best option is to stick to `np.dtype` since we do not have as descriptive and complete type as `np.dtype`. Couple routines like `get_dtype(arr: Tensor)` and `get_tensor_ty(dtype: np.dtype`) together with `np.dtype` solves all interfacing issues until nntile has its own `dtype`.

Comment (svtdanny) [reply wrappers/python/nntile/layer/batch_norm.py:56]: Your idea is clear and reasonable, at the same time It's very strange to require a user to use np.dtype to create nntile tensor. Second, we will have a lot numpy imports inside just for type usage

I tried to allow numpy and nntile types at the same type, it seems to be a reasonable compromise (see code)

File: wrappers/python/nntile/utils/constructors.py (hunk 15 above)
Comment (svtdanny) [reply wrappers/python/nntile/utils/constructors.py]: thanks! Took your code as baseline

File: wrappers/python/nntile/layer/batch_norm.py (hunk 16 above)
Comment (svtdanny) [reply wrappers/python/nntile/layer/batch_norm.py]: done

File: wrappers/python/nntile/layer/batch_norm.py (hunk 12 above)
Comment (svtdanny) [reply wrappers/python/nntile/layer/batch_norm.py]: done

File: wrappers/python/nntile/tensor.py (hunk 1 above)
Comment (svtdanny) [reply wrappers/python/nntile/tensor.py:820]: moved here: https://github.com/nntile/nntile/pull/57
I will rebase this later

File: wrappers/python/nntile/types.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    grad_required: bool
+
+    def __init__(self, value: TensorOrNone, grad: TensorOrNone,
+            grad_required: bool):
+        self.value = value
+        self.grad = grad
+        self.grad_required = grad_required
+
+    def __del__(self):
+        self.unregister()
```
[the remark this answers is not in the document]
Comment (svtdanny) [reply wrappers/python/nntile/types.py:28]: done in #57

File: wrappers/python/nntile/types.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
+from .nntile_core import tensor as core_tensor
+from .nntile_core.tensor import TensorTraits, Tensor_fp32, Tensor_fp64, \
+        Tensor_int64, Tensor_fp16, Tensor_bool
+from .nntile_core import TransOp, notrans, trans
+from typing import Union, List
+
+# Multiprecision tensor as a union type for all precisions
+Tensor = Union[core_tensor.Tensor_fp32, core_tensor.Tensor_fp64]
```
[the remark this answers is not in the document]
Comment (svtdanny) [reply wrappers/python/nntile/types.py:8]: done in #57

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 3 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py:97]: I don't see anybody use it here. Please let's move on it in different repo-wise pr

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 5 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py]: done

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 6 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py:39]: Its refactoring what we need to do globally. Please lets do in different PR

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 7 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py]: done

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 8 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py:56]: Its refactoring what we need to do globally. Please lets do in different PR

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 9 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py]: done

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 10 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py]: done

File: wrappers/python/tests/layer/test_batch_norm.py (hunk 11 above)
Comment (svtdanny) [reply wrappers/python/tests/layer/test_batch_norm.py]: done, but without pytest
Using pytest - is refactoring what we need to do globally. Please lets do in different PR

File: wrappers/python/nntile/utils/constructors.py (hunk 2 above)
Comment (svtdanny) [reply wrappers/python/nntile/utils/constructors.py]: Made "astensor" 
Just tensor will get confused with internal tensor and tensor.py
